messages/api.go

```diff
@@ -0,0 +1,170 @@
+// Copyright (c) 2019 NEC Laboratories Europe GmbH.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package messages defines interface for the protocol messages.
+//
+// Note that some interfaces are "sealed" with unexported methods.
+// This ensures the desired semantic of type assertions/switches with
+// those interfaces.
+//
+// To fulfill a sealed interface in another package, the new type
+// needs to explicitly embed a dedicated type that implements the
+// interface's unexported methods. For example, consider the following
+// definition of a sealed interface:
+//
+// 	package api
+//
+//	type Sealed interface {
+// 		DoSomething()
+// 		isSealed()
+// 	}
+//
+//	type IsSealed struct{}
+//
+//	func (IsSealed) isSealed() {}
```

Comment:
"Sealed interface" concept is new to me, but it sounds to strictly control visibility of the type, which should be good for code robustness. Anyway this patch only provides interfaces and I can't say anything about how each interface has each method for now.

Reply:
This concept is admittedly tricky, but seems to be the most reliable way to enforce explicit implementation of interfaces in Go (see https://github.com/golang/go/issues/34996).

Reply:
I was thinking _maybe_ combine those `type IsFoo struct{}` and their methods at the end of the file.

> Anyway this patch only provides interfaces and I can't say anything about how each interface has each method for now.

True, I just wanted to share the interfaces before I proceed with implementation. Now I have some implementation (not yet used), so I can share that, too.

Reply:
> This concept is admittedly tricky, but seems to be the most reliable way to enforce explicit implementation of interfaces in Go (see [golang/go#34996](https://github.com/golang/go/issues/34996)).

You're trying to improve Golang itself, wow 😲

Reply:
To have some fun at night :laughing: